### PR TITLE
Fix ManyToOneType.IsModified to handle both object instance and identifier passed to the "old" parameter 

### LINKED ISSUE #1496: Fix ManyToOneType.IsModified to handle both object instance and identifier passed to the parameter “old”.
# PROBLEM:
ManyToOneType.IsModified won't resolve identifier if object instance (not the identifier) was passed to the “old” parameter.
Similar issue https://hibernate.atlassian.net/browse/HHH-3730
 was reported in Hibernate with pull request https://github.com/hibernate/hibernate-orm/pull/728 

# DESCRIPTION:
The current code of IsModified method accepts 4 parameters:

`public override bool IsModified(object old, object current, bool[] checkable, ISessionImplementor session)`

It assumes that “old” is always Identifier and “current” is the object instance, so it tries to resolve Identifier for the “current” only: 

`return “GetIdentifierOrUniqueKeyType(session.Factory).IsDirty(old, GetIdentifier(current, session), session)”;`

SCENARIO 1: SelectBeforeUpdate

Typically, the method ManyToOneType.IsModified(…) is fired when flushing a session with an entity which contains a "many-to-one" property, but only when the "select-before-update" option is turned on for that class (and performed). For this scenario the identifier is always passed to “old” and it works fine.

SCENARIO 2: Event Listeners
It became common practice to do auditing with NHibernate Event Listeners and call Persister.FindModified() method: 

```
public class AuditEventListener : IPostUpdateEventListener
{
public void OnPostUpdate(PostUpdateEvent @event)
	{
		var modifiedItems = @event.Persister.FindModified(@event.OldState, @event.State, @event.Entity, @event.Session);
		// Process modified items.	
	}
}
```

For this scenario, @event.OldState and @event.State hold the object instances. It ends up going through calls: AbstractEntityPersister.FindModified(..) -> TypeHelper.FindModified(..) -> ManyToOneType. IsModified(..)
Eventually the object instances make into “old” parameter in ManyToOneType.IsModified(..). 
Once it happens, we have two problems:
1.	ManyToOneType.IsModified(..) returns **always true** comparing two ManyToOneType entities with regular (not composite) Ids.
2.	ManyToOneType.IsModified(..) throws **InvalidCastException** exception comparing two ManyToOneType entities with composite Ids

# SOLUTION:
We have to handle 4 possible scenarios for “old” 
1.	Id
2.	Composite Id
3.	Object Instance with Id
4.	Object Instance with Composite Id

A.	Leaving “old” as it is fails for (3) and (4)
B.	Resolving GetIdentifier(old, session) fails for (2). It was offered in https://github.com/hibernate/hibernate-orm/pull/728 and the code was pulled back because of composite id. In NHibernate NHibernate.Test.TypedManyToOne. TypedManyToOneTest will fail if we will do it this way.

### C.	I propose to check if “old” is Identifier before calling GetIdentifier(old, session):

*****************************************************************
OLD CODE (ManyToOneType)
*****************************************************************
```
*****************************************************************
public class ManyToOneType : EntityType
{
...
public override bool IsModified(object old, object current, bool[] checkable, ISessionImplementor session)
{
...
return GetIdentifierOrUniqueKeyType(session.Factory).IsDirty(old, GetIdentifier(current, session), session);
}
}
```

*****************************************************************
NEW FIXED CODE (ManyToOneType)
*****************************************************************
```
public class ManyToOneType : EntityType
{
...
public override bool IsModified(object old, object current, bool[] checkable, ISessionImplementor session)
{
...
var oldIdentifier = IsIdentifier(old, session) ? old : GetIdentifier(old, session);
	var currentIdentifier = GetIdentifier(current, session);
	return GetIdentifierOrUniqueKeyType(session.Factory).IsDirty(oldIdentifier, currentIdentifier, session);
}

private bool IsIdentifier(object value, ISessionImplementor session)
{
	var identifierType = GetIdentifierType(session);
	if (identifierType == null)
	{
		return false;
	}
	return value.GetType() == identifierType.ReturnedClass;
}
}	
```
*****************************************************************

### PR DESCRIPTION
**Pull request by mishaurubkov**
… identifier passed to the parameter “old”.

The tests `EventListener_Entity_ChangeProperty` and `EventListener_EntityWithCompositeId_ChangeProperty` fail for current code in `ManyToOneType.IsModified(…)` where `old` parameter is not parsed.

The tests `SelectBeforeUpdate_EntityWithCompositeId_ChangeProperty` and `SelectBeforeUpdate_ManyToOneWithCompositeId_ChangeProperty` fail
 if we apply Hibernate fix [HHH-3730](https://hibernate.atlassian.net/browse/HHH-3730) `(var oldIdentifier =GetIdentifier(old, session))`

All tests pass if we apply proposed fix: `var oldIdentifier = IsIdentifier(old, session) ? old : GetIdentifier(old, session);`

I added extra tests to see what data hit the `ManyToOneType.IsModified(…)`. We can leave just 4 tests listed above and remove or set to `[Ignore]` others.

`ManyToOneType.IsModifiedAsync(...)` could be fixed same way but would like to hear the feedback on the issue and proposed solution first.

Fixes #1496.

### CODE REVIEW

**Comment by fredericDelaporte** [comment]
> ManyToOneType.IsModifiedAsync(...) could be fixed same way but would like to hear the feedback on the issue and proposed solution first.

No need to wait, async code is generated: just regenerate it. (Do not manually edit it.) Follow [contributing](https://github.com/nhibernate/nhibernate-core/blob/master/CONTRIBUTING.md) instructions. [Async generation](https://github.com/nhibernate/nhibernate-core/blob/master/CONTRIBUTING.md#regenerate-async-code) is an option of the [build menu](https://github.com/nhibernate/nhibernate-core/blob/master/CONTRIBUTING.md#the-build-menu).

**Comment by mishaurubkov** [comment]
DONE. Generated Async method with tests. See #98b1918 commit.

File: src/NHibernate.Test/NHSpecificTest/GH1496/AuditEventListener.cs [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+			ModifiedItems.Clear();
+			isActive = true;
+		}
+
+		public void Stop()
+		{
+			isActive = false;
+		}
+
+		public async Task OnPostUpdateAsync(PostUpdateEvent @event, CancellationToken cancellationToken)
```
Comment (fredericDelaporte) [inline src/NHibernate.Test/NHSpecificTest/GH1496/AuditEventListener.cs]: Normally there is a no need to write this async method too. The async generator will do it for you too. Of course this means that at first your class will lack some interface method implementations and the project will not compile till you run the async generator, but this is the preferred way.

File: src/NHibernate.Test/NHSpecificTest/GH1496/Entities.cs [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+
+		public Person(int id, string name, Address address)
+		{
+			Id = id;
+			Name = name;
+			Address = address;
+		}
+	}
+
+	// Entity with ManyToOne type (Contact) with comnposite (ContactId) Id.
```
Comment (fredericDelaporte) [inline src/NHibernate.Test/NHSpecificTest/GH1496/Entities.cs]: Typo on `comnposite`.

File: src/NHibernate.Test/NHSpecificTest/GH1496/Entities.cs [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+
+	[Serializable]
+	public class Contact
+	{
+		public virtual ContactIdentifier ContactIdentifier { get; set; }
+		public virtual string Phone { get; set; }
+	}
+
+
+	public class ContactIdentifier : IEquatable<ContactIdentifier>
```
Comment (fredericDelaporte) [inline src/NHibernate.Test/NHSpecificTest/GH1496/Entities.cs]: Shouldn't this be marked serializable too?

File: src/NHibernate.Test/NHSpecificTest/GH1496/Entities.cs [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+		{
+			return !Equals(left, right);
+		}
+
+
+
+	}
+
+
+
```
Comment (fredericDelaporte) [inline src/NHibernate.Test/NHSpecificTest/GH1496/Entities.cs]: Please clean the file from such consecutive blank lines.

**Comment by mishaurubkov** [comment]
I did cleanup and committed the changes:
1) OnPostUpdateAsync was created using generator
2) Removed spaces
3) I didn't really need [Serializable] so I removed it.

**Pull request merged**